vfs/vfscache/downloaders/downloaders.go

```diff
@@ -576,9 +584,15 @@ func (dl *downloader) stopAndClose(inErr error) (err error) {
 func (dl *downloader) download() (n int64, err error) {
 	// defer log.Trace(dl.dls.src, "")("err=%v", &err)
 	n, err = dl.in.WriteTo(dl)
-	if err != nil && errors.Cause(err) != asyncreader.ErrorStreamAbandoned {
-		return n, errors.Wrap(err, "vfs reader: failed to write to cache file")
+	if err != nil {
+		if fserrors.IsErrNoSpace(err) {
```

Comment:
I'm not sure you need this bit of code at all - `IsErrNoSpace()` should be able to unwrap the error just fine so I don't see the purpose?

Reply:
The original code does not return non-nil error (or the ENOSPC error that we are considering currently) unless errors.Cause(err) != asyncreader.ErrorStreamAbandoned.
```
if err != nil && errors.Cause(err) != asyncreader.ErrorStreamAbandoned {
    return n, errors.Wrap(err, "vfs reader: failed to write to cache file")
}
return n, nil
```
Maybe that's bug?  Or I am missing something.
Should the line be 
```
return n, err
```

Reply:
The original code says in English, return an error unless it is ErrorStreamAbandoned in which case don't.

Your new code does exactly the same thing. However I think the original code would work just fine if `err == ENOSPC` then it would be returned in this line

    return n, errors.Wrap(err, "vfs reader: failed to write to cache file")

So I think the major difference between your code and the original is that the ENOSPC error doesn't get wrapped in the `errors.Wrap()` call. I don't think that matters since we are unwrapping the errors to find the ENOSPC anyway so one more layer of wrap doesn't make any difference.

So I think your change works fine, it just isn't needed.

Reply:
You are right!  I miss-read the second part of the if condition as "==" (instead of !=).
if err != nil && errors.Cause(err) **==** asyncreader.ErrorStreamAbandoned {